python-threatexchange/threatexchange/exchanges/impl/ncmec_api.py

```diff
@@ -37,23 +37,45 @@ class NCMECCheckpoint(
     NCMEC IDs seem to stay around forever, so no need for is_stale()
     """
 
+    PAGING_URL_EXPIRATION = 12 * 60 * 60
+
     # The biggest value of "to", and the next "from"
     get_entries_max_ts: int
+    # A url to fetch the next page of results
+    # Only reference this value through get_paging_url_if_recent
+    paging_url: str = ""
+    # a timestamp for the last fetch time, specifically used with a pagingpyth_url
+    # NCMEC suggests not storing paging_urls long term so we consider them invalid
+    # 12hr after the last_fetch_time
+    last_fetch_time: int = field(hash=True, default_factory=lambda: int(time.time()))
 
     def get_progress_timestamp(self) -> t.Optional[int]:
         return self.get_entries_max_ts
 
+    def get_paging_url_if_recent(self) -> str:
+        if int(time.time()) - self.last_fetch_time < self.PAGING_URL_EXPIRATION:
+            return self.paging_url
+        return ""
+
     @classmethod
     def from_ncmec_fetch(cls, response: api.GetEntriesResponse) -> "NCMECCheckpoint":
         """Synthesizes a checkpoint from the API response"""
-        return cls(response.max_timestamp)
+        return cls(response.max_timestamp, response.next, int(time.time()))
 
     def __setstate__(self, d: t.Dict[str, t.Any]) -> None:
         """Implemented for pickle version compatibility."""
         # 0.99.0 => 1.0.0:
         ### field 'max_timestamp' renamed to 'get_entries_max_ts'
         if "max_timestamp" in d:
             d["get_entries_max_ts"] = d.pop("max_timestamp")
+
+        # 1.0.0 => 1.2.3:
+        # Add last_fetch_time
+        # note: the default_factory value was not being set correctly when
+        # reading from pickle
+        if not "last_fetch_time" in d:
+            d["last_fetch_time"] = int(time.time())
```

Comment:
I was getting `AttributeError: 'NCMECCheckpoint' object has no attribute 'last_fetch_time'` without this in the `test_state_compatibility` test

seems sort of related to https://github.com/pydantic/pydantic/issues/7821, since `default` was working (but wouldn't work if we want to set it to the current time)

Reply:
Wow, tough bug, good find on the fix